File: Algo/Leetcode/416PartitionEqualToSubsetSum.cpp
// dp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum = 0;
        for(auto c:nums) sum+=c;
        if(sum%2) return false;
        sum /= 2;

        int n = nums.size();
        vector<bool> dp(sum+1);
        dp[0] = true;
        for(auto i:nums){
            for(int j=sum; j>=i; j--)
                dp[j] = dp[j] || dp[j-i];
        }
        return dp[sum];
    }
};

File: Algo/Leetcode/416PartitionEqualToSubsetSum.py
class Solution(object):
    def canPartition(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        s = sum(nums)
        n = len(nums)
        if s % 2 != 0:
            return False
        s = s / 2

        dp = [[False for _ in range(s+1)] for _ in range(n+1)] # (n+1, s+1)
        for i in range(n+1):
            dp[i][0] = True     # without any can be sum 0

        for i in range(1, n+1):
            for k in range(1, s+1):
                dp[i][k] = (dp[i-1][k-nums[i-1]] if nums[i-1] <= k else False) or dp[i-1][k]

        return dp[-1][-1]